File: basic_dsa_two_hundred_seventeen.cpp
/**
 * Basic DSA - Find Peak Element (LeetCode 162)
 * Find a peak element (greater than neighbors) in O(log n).
 */
#include <iostream>
#include <vector>

int findPeakElement(std::vector<int>& nums) {
    int lo = 0, hi = (int)nums.size() - 1;
    while (lo < hi) {
        int mid = lo + (hi - lo) / 2;
        if (nums[mid] > nums[mid + 1]) hi = mid;
        else lo = mid + 1;
    }
    return lo;
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Find Peak Element\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> a1 = {1, 2, 3, 1};
    std::cout << "[1,2,3,1] -> index " << findPeakElement(a1) << "\n";

    std::vector<int> a2 = {1, 2, 1, 3, 5, 6, 4};
    std::cout << "[1,2,1,3,5,6,4] -> index " << findPeakElement(a2) << "\n";
}

File: basic_dsa_two_hundred_seventeen.py
"""
Basic DSA - Find All Duplicates in an Array
Given an integer array nums of length n where all integers are in the range [1, n]
and each integer appears once or twice, return all the integers that appear twice.
"""


def find_duplicates(nums):
    """
    Mark presence by negating nums[abs(x) - 1]; if already negative, it's a duplicate.
    Time Complexity: O(n)
    Space Complexity: O(1) extra (output excluded)
    """
    res = []
    for x in nums:
        idx = abs(x) - 1
        if nums[idx] < 0:
            res.append(abs(x))
        else:
            nums[idx] = -nums[idx]
    return res


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Find All Duplicates in an Array")
    print("=" * 60)

    print(sorted(find_duplicates([4, 3, 2, 7, 8, 2, 3, 1])))  # [2, 3]
    print(sorted(find_duplicates([1, 1, 2])))  # [1]
